### ISSUE TITLE
array_filter may return an empty-array, not detected by psalm

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#4977]
https://psalm.dev/r/36ad83e104

**Comment by orklah** [comment]
If you want to give it a try, the fix should be in ArrayFilterReturnTypeProvider.php .

It's the file that returns the return type for array_filter depending on its params.

Otherwise, I'll take a look this afternoon

**Comment by orklah** [comment]
This is more complicated than I expected. I found two issues:
- Your case seem to be related to the concatenation in the offset. It's fixed when we extract the concatenation: https://psalm.dev/r/8a7f2b4d83 . Not sure how to fix this yet
- When passing a Keyed Array to array_filter with a custom function in second param, Psalm doesn't keep Keyed array structure: https://psalm.dev/r/1f71a173d0 . I found a way to fix this, but in the bad way, as it makes the solution I posted above invalid.

I'll keep this on the side for now, it seems to have something to do with assertions and it's a system I didn't touch yet.

**Comment by muglug** [comment]
@danog could you provide a more real-world example?

**Comment by danog** [comment]
Hmmm, this actually is a real-world example, but I'll try simplifying it :) 

I noticed, too that the issue is only present if the key name is a concatenation with a variable, that's why I couldn't initially find a way to further simplify this use case.

**Comment by muglug** [comment]
The problem can be simplified to 

```php
<?php
$data = ['val_test' => ['test' => 'other']];

if (!empty($data['val_' . 'test'])) {
	$data['val_' . 'test'] = array_filter($data['val_' . 'test'], function ($_) {
		return false;
	});
	if (empty($data['val_' . 'test'])) {}
}
```

but you can easily work around this by removing the unnecessary concatenation:

```php
<?php
$data = ['val_test' => ['test' => 'other']];

if (!empty($data['val_test'])) {
	$data['val_test'] = array_filter($data['val_test'], function ($_) {
		return false;
	});
	if (empty($data['val_test'])) {}
}
```

**Comment by danog** [comment]
Hmm, but this does not fix the issue at all; in my case, the issue is triggered exactly when the key is a concatenation of a string and a variable, and still triggers false positives during psalm analysis.

**Comment by orklah** [comment]
Well, your code was supposed to be the real life example and the concatenation is useless in the example.

Also, please note that if you concatenate in a variable and then use the variable, the issue disappear https://psalm.dev/r/8a7f2b4d83
